### PR TITLE
New fields added for diagnostics

### PR DESCRIPTION
**Pull request by skandalaCLIMA**
# Description

A clear and concise description of the code with usage.

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and run `julia .dev/format.jl`

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Diagnostics/diagnostic_fields.jl [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+using CUDAapi
+using CuArrays
+using StaticArrays
+using KernelAbstractions
+using CLIMA.Mesh.Geometry
+
+const _ξ1x1, _ξ2x1, _ξ3x1 = Grids._ξ1x1, Grids._ξ2x1, Grids._ξ3x1
+const _ξ1x2, _ξ2x2, _ξ3x2 = Grids._ξ1x2, Grids._ξ2x2, Grids._ξ3x2
+const _ξ1x3, _ξ2x3, _ξ3x3 = Grids._ξ1x3, Grids._ξ2x3, Grids._ξ3x3
+const _ρ, _ρu, _ρv, _ρw = 1, 2, 3, 4
```
Comment (charleskawczynski) [inline src/Diagnostics/diagnostic_fields.jl:12]: Shouldn't we use `vars_state` instead of hard-coded indexes?

Comment (skandalaCLIMA) [reply src/Diagnostics/diagnostic_fields.jl:12]: This is a first rough draft for backup purposes. Removing hard-code indexes is on the TODO list.

File: src/Diagnostics/diagnostic_fields.jl [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+"""
+struct SphericalCoord{FT <: AbstractFloat, FTA2D <: AbstractArray{FT, 2}}
+    "radius"
+    r::FTA2D
+    "latitude/polar angle"
+    θ::FTA2D
+    "longitude/azimuthal angle"
+    ϕ::FTA2D
+
```
Comment (simonbyrne) [inline src/Diagnostics/diagnostic_fields.jl:181]:
```suggestion
struct SphericalCoord{FT <: AbstractFloat, FTA3D <: AbstractArray{FT, 2}, FTA2D <: AbstractArray{FT, 2}}
    data::FTA3D
    "radius"
    r::FTA2D
    "latitude/polar angle"
    θ::FTA2D
    "longitude/azimuthal angle"
    ϕ::FTA2D

```

File: src/Diagnostics/diagnostic_fields.jl [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    ∂ξ₃∂r::FTA2D
+    "∂ξ₃∂θ"
+    ∂ξ₃∂θ::FTA2D
+    "∂ξ₃∂ϕ"
+    ∂ξ₃∂ϕ::FTA2D
+
+    function SphericalCoord(Npl, Nel, ::Type{FT}) where {FT <: AbstractFloat}
+        DA = CLIMA.array_type()
+        return new{FT, typeof(DA{FT}(undef, Npl, Nel))}(
+            DA{FT}(undef, Npl, Nel),
```
Comment (simonbyrne) [inline src/Diagnostics/diagnostic_fields.jl:213]:
```suggestion
        DA = CLIMA.array_type()
        data = DA{FT}(undef, Npl, Nfields, Nel)
        r = @view(data[:,1,:])
        ...
        return new{FT,typeof(data), typeof(r)}(
            data, r, ...
```

File: src/Diagnostics/diagnostic_fields.jl [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+using CUDAnative
+using CUDAapi
+using CuArrays
+using StaticArrays
+using KernelAbstractions
+using CLIMA.Mesh.Geometry
+
+const _ξ1x1, _ξ2x1, _ξ3x1 = Grids._ξ1x1, Grids._ξ2x1, Grids._ξ3x1
+const _ξ1x2, _ξ2x2, _ξ3x2 = Grids._ξ1x2, Grids._ξ2x2, Grids._ξ3x2
+const _ξ1x3, _ξ2x3, _ξ3x3 = Grids._ξ1x3, Grids._ξ2x3, Grids._ξ3x3
```
Comment (simonbyrne) [inline src/Diagnostics/diagnostic_fields.jl:12]:
```suggestion
import Grids: _ξ1x1, _ξ2x1, _ξ3x1, ...
```

File: src/Diagnostics/diagnostic_fields.jl (hunk 1 above)
Comment (simonbyrne) [reply src/Diagnostics/diagnostic_fields.jl:13]: see if you can use `varsindex` for this?

Comment (simonbyrne) [reply src/Diagnostics/diagnostic_fields.jl:13]: https://github.com/climate-machine/CLIMA/blob/0ca53bec15528e1b62814307611212907481513e/test/Utilities/VariableTemplates/varsindex.jl

Comment (simonbyrne) [reply src/Diagnostics/diagnostic_fields.jl:13]: https://github.com/climate-machine/CLIMA/blob/9015194811b1d3edd75bb2dc0601ac7e70912e4c/examples/Microphysics/ex_1_saturation_adjustment.jl#L242-L247

File: src/Diagnostics/diagnostic_fields_kernels.jl [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+
+    for k in 1:qm1
+        ijk = i + ((j-1) + (k-1)*qm1)*qm1
+
+        x = vgeo[ijk, _x, e]
+        y = vgeo[ijk, _y, e]
+        z = vgeo[ijk, _z, e]
+
+        r[ijk,e] = sqrt(x*x + y*y + z*z)
+        θ[ijk,e] = atan(sqrt(x*x + y*y)/z)
```
Comment (simonbyrne) [inline src/Diagnostics/diagnostic_fields_kernels.jl:187]:
```suggestion
        θ[ijk,e] = atan(sqrt(x*x + y*y), z)
```

File: test/Diagnostics/diagnostic_fields_LES.jl [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        CLIMA.Mesh.Grids.vgeoid.x3id
+
+        Q = solver_config.Q
+        dg = solver_config.dg
+        grid = dg.grid
+        vgeo = grid.vgeo
+        Nel = size(Q.realdata, 3)
+        Npl = size(Q.realdata, 1)
+        println("typeof(vgeo) = ", typeof(vgeo))
+        println("size(vgeo) = ", size(vgeo))
```
Comment (simonbyrne) [inline test/Diagnostics/diagnostic_fields_LES.jl:171]: remove `print` statements

File: src/Diagnostics/diagnostic_fields.jl [hunk 7]
```diff
@@ -0,0 +1,318 @@
+using DocStringExtensions
+using CUDAdrv
+using CUDAnative
+using CUDAapi
+using CuArrays
```
Comment (simonbyrne) [inline src/Diagnostics/diagnostic_fields.jl:5]: Do we still need to load these?

**Comment by simonbyrne** [comment]
@skandalaCLIMA What else needs to be done here?

**Pull request merged**